src/Codeception/Lib/InnerBrowser.php

```diff
@@ -996,35 +968,45 @@ public function grabValueFrom($field)
         if (!$nodes->count()) {
             throw new ElementNotFound($field, 'Field');
         }
+        return $this->proceedGetValueFromField($nodes);
+    }
 
+    /**
+     * @param Crawler $nodes
+     * @return array|mixed|string
+     */
+    protected function proceedGetValueFromField(Crawler $nodes)
+    {
+        $values = [];
         if ($nodes->filter('textarea')->count()) {
-            return $nodes->filter('textarea')->text();
+            return (new TextareaFormField($nodes->filter('textarea')->getNode(0)))->getValue();
         }
+
         if ($nodes->filter('input')->count()) {
-            return $nodes->filter('input')->attr('value');
+            $input = $nodes->filter('input');
+            if ($input->attr('type') == 'checkbox' or $input->attr('type') == 'radio') {
+                $values = [];
+                $input = $nodes->filter('input:checked');
+                foreach ($input as $checkbox) {
+                    $values[] = $checkbox->getAttribute('value');
+                }
+                return $values;
+            }
+            return (new InputFormField($nodes->filter('input')->getNode(0)))->getValue();
         }
-
         if ($nodes->filter('select')->count()) {
-            /** @var  \Symfony\Component\DomCrawler\Crawler $select */
-            $select      = $nodes->filter('select');
-            $isMultiple  = (bool)$select->attr('multiple');
-            $results     = [];
-            foreach ($select->children() as $option) {
-                /** @var  \DOMElement $option */
-                if ($option->getAttribute('selected') == 'selected') {
-                    $val = $option->getAttribute('value');
-                    if (!$isMultiple) {
-                        return $val;
-                    }
-                    $results[] = $val;
-                }
+            $field = new ChoiceFormField($nodes->filter('select')->getNode(0));
+            $options = $nodes->filter('option[selected]');
+            foreach ($options as $option) {
+                $values[] = $option->getAttribute('value');
```

Comment:
the strange thing that I couldn't use `$field->getValue()` here as it returned all options no matter if they were selected or not :(